text/0000-bom-in-layer-metadata.md

```diff
@@ -0,0 +1,108 @@
+
+```
+[[bom]]
+name = "<dependency name>"
+
+[bom.metadata]
+# arbitrary metadata describing the dependency
+```
+
+For each dependency contributed to the app image or build environment, the buildpack layer:
+
+- SHOULD add a bill-of-materials entry to the bom array describing the dependency, where:
+  - name is REQUIRED.
+  - metadata MAY contain additional data describing the dependency.
+  
+The buildpack SHOULD NOT add bom to a layer content metadata file describing the contents that are not contributed by the same layer. `BOM` entries that are not related to a specific layer should go in the buildpack level `BOM` entries in `launch.toml` or `build.toml`.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle would be responsible for the generation of the appropriate metadata files and labels given partial `BOM` entries in the `launch.toml`, `build.toml` and `<layer-name>.toml`. The generation of these artifacts given `BOM` entries in `<layer-name>.toml` would be controlled as follows - 
+
+- If the layer has `types.launch` set to `true` then all the `bom` entries will be contributed to the app image in `io.buildpacks.build.metadata` and `<layers>/config/metadata.toml`.
+- If the layer has `types.build` set to `true` then all the `bom` entries will be written to `report.toml`.
+- If the layer has `types.cache` set to `true` and if the layer is reused in the next build, the respective `bom` entries should be re-populated from the cached layer content metadata file.
+- If `types.cache` is set to `false`, but `types.launch` is set to `true` before a rebuild, the layer content metdata file should be restored with the `BOM` section. If the buildpack sets `types.launch` to `true` after the re-build, the final value of `bom` after the re-build should be used.
+- If neither `types.launch` or `types.build` are set to `true` the `bom` entries will not be exported anywhere. However they will be preserved in the layer metadata file if `types.cache` is set to `true`.
+
+All `bom` entries should respect any combination of the above.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+Additional complexity and utility pushed down to the lifecycle.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Keep it as it is and provide a similar functionality in language bindings like `libcnb`. Currently the layer content metadata files are already being used to store and regenerate the `BOM` in buildpacks from the `paketo` project.
+
+# Prior Art
+[prior-art]: #prior-art
+
+<!-- TODO -->
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+N/A
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+A new array of tables `[[bom]]` to be added to the layer content metadata files  - 
+
+```toml
+[types]
+  launch = false
+  build = false
+  cache = false
+
+[metadata]
+# buildpack-specific data
+
+[[bom]]
+name = "<string>"
+
+[bom.metadata]
+# Additional metadata
+```
```

Comment:
I tried to make this a suggestion but couldn't figure out how to put a code block in a suggestion...


A new `layer` key should be added to the `bom` entries in `io.buildpacks.build.metadata` - 
```javascript
{
  "processes": [
    {
      "type": "<process-type>",
      "command": "<command>",
      "args": [
        "<args>"
      ],
      "direct": false
    }
  ],
  "buildpacks": [
    {
      "id": "<buildpack ID>",
      "version": "<buildpack version>",
      "homepage": "<buildpack homepage>"
    }
  ],
  "bom": [
    {
      "name": "<bom-entry-name>",
      "layer": "<layer-name>",
      "metadata": {
        // arbitrary buildpack provided metadata
      },
      "buildpack": {
        "id": "<buildpack ID>",
        "version": "<buildpack version>"
      }
    },
  ],
 "launcher": {
    "version": "<launcher-version>",
    "source": {
      "git": {
        "repository": "<launcher-source-repository>",
        "commit": "<launcher-source-commit>"
      }
    }
  }
}
```


Paketo buildpack add the layer name to the BOM entries so consumers of the BOM know which layer contains what. Do we want to spec this data now that it is available? For entries that are not associated with a layer we could:
1.  omit the `layer` key
2. assume it is in the `app` layer. The drawback here is that a buildpack could theoretically be adding BOM entries describing the stack image itself? But I think we would rather provide a mechanism that allows the stack to BYOBOM entries instead of encouraging this pattern.

Reply:
Went with option 1